sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/StatusDetection.java

```diff
@@ -19,34 +19,64 @@
  */
 package org.sonar.scanner.scan.filesystem;
 
+import javax.annotation.CheckForNull;
 import javax.annotation.concurrent.Immutable;
-
 import org.apache.commons.lang.StringUtils;
 import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.batch.fs.internal.DefaultInputFile;
 import org.sonar.scanner.repository.FileData;
 import org.sonar.scanner.repository.ProjectRepositories;
+import org.sonar.scanner.scm.ScmChangedFiles;
+
+import static org.sonar.api.batch.fs.InputFile.Status.*;
 
 @Immutable
-class StatusDetection {
+public class StatusDetection {
 
   private final ProjectRepositories projectRepositories;
+  private final ScmChangedFiles scmChangedFiles;
 
-  StatusDetection(ProjectRepositories projectSettings) {
+  public StatusDetection(ProjectRepositories projectSettings, ScmChangedFiles scmChangedFiles) {
     this.projectRepositories = projectSettings;
+    this.scmChangedFiles = scmChangedFiles;
   }
 
-  InputFile.Status status(String projectKeyWithBranch, String relativePath, String hash) {
-    FileData fileDataPerPath = projectRepositories.fileData(projectKeyWithBranch, relativePath);
+  InputFile.Status status(String projectKeyWithBranch, DefaultInputFile inputFile, String hash) {
+    FileData fileDataPerPath = projectRepositories.fileData(projectKeyWithBranch, inputFile.relativePath());
     if (fileDataPerPath == null) {
-      return InputFile.Status.ADDED;
+      return checkChanged(ADDED, inputFile);
     }
     String previousHash = fileDataPerPath.hash();
     if (StringUtils.equals(hash, previousHash)) {
-      return InputFile.Status.SAME;
+      return SAME;
+    }
+    if (StringUtils.isEmpty(previousHash)) {
+      return checkChanged(ADDED, inputFile);
+    }
+    return checkChanged(CHANGED, inputFile);
+  }
+
+  /**
+   * If possible, get the status of the provided file without initializing metadata of the file.
+   * @return null if it was not possible to get the status without calculating metadata
+   */
+  @CheckForNull
+  public InputFile.Status getStatusWithoutMetadata(String projectKeyWithBranch, DefaultInputFile inputFile) {
+    FileData fileDataPerPath = projectRepositories.fileData(projectKeyWithBranch, inputFile.relativePath());
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "relativePath"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)